foundry/auto_self_nominate.rs

```diff
@@ -0,0 +1,157 @@
+use codechain_types::transaction::{Action, Transaction};
+use primitives::Bytes;
+use std::sync::Arc;
+use std::thread;
+use std::time::Duration;
+
+pub struct AutoSelfNomination {
+    client: Arc<dyn ConsensusClient>,
+    signer: EngineSigner,
+}
+
+impl AutoSelfNomination {
+    pub fn new(client: Arc<dyn ConsensusClient>) -> Arc<Self> {
+        Arc::new(Self {
+            client,
+            signer: Default::default(),
+        })
+    }
+
+    pub fn send_self_nominate_transaction(&self, matches: &ArgMatches) {
+        let config = load_config(matches).unwrap();
+        let account_address = config.mining.engine_signer.unwrap();
+        let passpharase = config.operating.password_path.unwrap();
+        let defualt_metadata = config.mining.self_nomination_metadata.unwrap();
+        let target_deposite = config.mining.self_target_deposit.unwrap();
+        let interval = config.mining.self_nomination_interval.unwrap();
+        let self_client = self.client.clone();
+
+        let self_signer = self.signer.clone();
+        thread::Builder::new()
+            .name("Auto Self Nomination".to_string())
+            .spawn(move || loop {
+                AutoSelfNomination::send(
+                    &self_client,
+                    &self_signer,
+                    &account_address,
+                    &passpharase,
+                    &defualt_metadata,
+                    target_deposite,
+                );
+                thread::sleep(Duration::from_millis(interval));
+            })
+            .unwrap();
+    }
+
+    fn send(
+        client: &Arc<dyn ConsensusClient>,
+        signer: &EngineSigner,
+        account_address: &PlatformAddress,
+        passphrase: &String,
+        metadata: &String,
+        targetdep: u64,
+    ) {
+        let metabytes = metadata.rlp_bytes();
+        let mut dep = targetdep;
+        let address = account_address.address();
+        let block_id = BlockId::Latest;
+        let state = client.state_at(block_id).unwrap();
+
+        let banned = Banned::load_from_state(&state).unwrap();
+        if banned.is_banned(address) {
+            cerror!(ENGINE, "Account is banned");
+            return
+        }
+
+        let jailed = Jail::load_from_state(&state).unwrap();
+        let prisoner = jailed.get_prisoner(&address).unwrap();
+        let current_term = client.current_term_id(block_id).unwrap();
+        if prisoner.custody_until <= current_term {
+            cerror!(ENGINE, "Account is still in custody");
+        }
+
+        let candidate = Candidates::load_from_state(&state).unwrap();
+        let candidate_need_nomination = candidate.get_candidate(&address).unwrap();
+        if !candidate_need_nomination.nomination_ends_at.lt(&current_term) {
+            cerror!(ENGINE, "No need self nominate");
+        }
+        if !candidate_need_nomination.nomination_ends_at.min(current_term).lt(&current_term) {
+            cerror!(ENGINE, "No need self nominate");
+        }
```

Comment:
It seems that the condition is wrong.
We should send self-nomination transaction when `nomination_ends_at <= current_term + SOME_CONSTANT`

The typescript version "auto-self-nominate" script is using `NEED_NOMINATION_UNDER_TERM_LEFT` environment variable instead of `SOME_CONSTANT`.